src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs

```diff
@@ -344,141 +354,154 @@ private static unsafe bool AllowHyphenDuringParsing(NumberFormatInfo info)
-                        }
-                        else if (ch == '(' && ((options & NumberStyles.AllowParentheses) != 0) && ((state & StateSign) == 0))
-                        {
-                            state |= StateSign | StateParens;
-                            number.sign = true;
-                        }
-                        else if (currSymbol != null && (next = MatchChars(p, strEnd, currSymbol)) != null)
-                        {
-                            state |= StateCurrency;
-                            currSymbol = null;
+                    returnValue = new FormattingSpecification(numfmt.CurrencySymbol,
+                        numfmt.CurrencyDecimalSeparator,
+                        numfmt.CurrencyGroupSeparator);
+                }
+                else
+                {
+                    returnValue = new FormattingSpecification(null,
+                        numfmt.NumberDecimalSeparator,
+                        numfmt.NumberGroupSeparator);
+                }
 
-                            // We already found the currency symbol. There should not be more currency symbols. Set
-                            // currSymbol to NULL so that we won't search it again in the later code path.
-                            p = next - 1;
-                        }
-                        else
-                        {
-                            break;
-                        }
-                    }
-                    ch = ++p < strEnd ? *p : '\0';
+                return returnValue;
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static unsafe char* MatchNegativeSignChars(char* p, char* pEnd, string negativeSign, bool allowHyphenDuringParsing)
+            {
+                char *ret = MatchChars(p, pEnd, negativeSign);
+                if (ret == null && allowHyphenDuringParsing && p < pEnd && *p == '-')
+                {
+                    ret = p + 1;
                 }
 
-                int digCount = 0;
-                int digEnd = 0;
+                return ret;
+            }
+
+            private static unsafe StringBuilder ParseCommon(
+                char* strEnd,
+                NumberStyles options,
+                ref NumberBuffer number,
+                NumberFormatInfo numfmt,
+                FormattingSpecification formattingSpecification,
+                ref NumberStates state,
+                ref char* p,
+                ref char ch,
+                ref char* next)
+            {
+                StringBuilder returnValue;
+
+                returnValue = new StringBuilder(MaxDigitsCount);
                 while (true)
                 {
-                    if ((ch >= '0' && ch <= '9') || (((options & NumberStyles.AllowHexSpecifier) != 0) && ((ch >= 'a' && ch <= 'f') || (ch >= 'A' && ch <= 'F'))))
+                    if (IsDecimalChar(ch) || ((options.HasFlag(NumberStyles.AllowHexSpecifier)) && IsHexadecimalChar(ch)))
```

Comment:
Nit: Extra parens around `options.Has...` can be removed. Normally I wouldn't care about it, but in this case, I had to re-review the `&&` condition becuase the parens threw me off